cmd/cloudshell_open/build.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+)
+
+// buildContainer will use docker to build a container if a Dockerfile is
+// available. Otherwise it will fallback on trying to use buildpacks.
+func buildContainer(dir, image string) error {
+	if dockerFileExists(dir) {
+		if err := dockerBuild(dir, image); err != nil {
+			return err
+		}
+
+		if err := dockerPush(image); err != nil {
+			return fmt.Errorf("failed to push image to %s: %+v", image, err)
+		}
+
+		return nil
+	}
+
+	// Use buildpacks
+	return packBuildAndPublish(dir, image)
```

Comment:
See my comment in the relevant issue. 
I think we should not fall back, we should try this path ONLY if a particular file exists. 

Similarly the error shouldnt say “pack failed” IF there’s no dockerfile in dir. it should say “no dockerfile or procfile”. So i think you need a procfileExists() here.

Reply:
Procfile isn't required for buildpacks.  They are really designed to have nothing special in the repo to make them buildpack compatible.  So I think it makes sense to fallback to buildpacks since buildpacks have a detect phase which will fail if it can't build the repo.

Reply:
But the primary path is dockerfile. We don’t wanna fail with buildpack error and totally don’t mention dockerfile when fallback fails, too. 

My experimentation with heroku buildpacks showed me Procfile is required. Is this totally incorrect?

Reply:
Since I don't know Go I can't totally tell but I think the experience should be something like:
```
if Dockerfile.exists then doDockerBuild # if this fails stop and display the error
else echo "Dockerfile not found, trying buildpacks"; doBuildpackBuild # if this fails stop and display the error
```

The `Procfile` can often be inferred by the buildpack.  But in cases when it can't, or when something custom is needed, it can be provided as an override.

Reply:
@ahmetb I got this working with Go for https://github.com/poy/knative-testing